### PR TITLE
OCL version of Detection output layer enabled

### PR DESCRIPTION
**Pull request by listenlink**
Hi @gongzg ,

This patch enabled ocl version of detection_output_layer and pass the corresponding test cases.

### CODE REVIEW

File: src/caffe/util/bbox_util.cu [hunk 1]
```diff
@@ -232,16 +271,24 @@ __global__ void PermuteDataKernel(const int nthreads,
...
+  if (Caffe::GetDefaultDevice()->backend() == BACKEND_CUDA) {
+#ifdef USE_CUDA
+    // NOLINT_NEXT_LINE(whitespace/operators)
+    PermuteDataKernel<Dtype><<<CAFFE_GET_BLOCKS(nthreads),
+        CAFFE_CUDA_NUM_THREADS>>>(nthreads, data, num_classes, num_data,
+        num_dim, new_data);
+    CUDA_POST_KERNEL_CHECK;
+#endif //USE_CUDA
+  } else {
+#ifdef USE_GREENTEA
```
Comment (gongzg) [inline src/caffe/util/bbox_util.cu]: Please add a NOT_IMPLEMENTED stub in this type of code block.

File: src/caffe/util/bbox_util.cu [hunk 2]
```diff
@@ -319,39 +367,47 @@ void SoftMaxGPU(const Dtype* data, const int outer_num,
...
+        scale_data);
+    // divide
+    // NOLINT_NEXT_LINE(whitespace/operators)
+    kernel_channel_div<Dtype><<<CAFFE_GET_BLOCKS(count),
+        CAFFE_CUDA_NUM_THREADS>>>(count, outer_num, channels, inner_num,
+        scale_data, prob);
+#endif //USE_CUDA
+  } else {
+#ifdef USE_GREENTEA
+#endif //USE_GREENTEA
```
Comment (gongzg) [inline src/caffe/util/bbox_util.cu]: Please add a NOT_IMPLEMENTED stub in this type of code block.

File: src/caffe/util/bbox_util.cu [hunk 3]
```diff
@@ -375,16 +431,24 @@ __global__ void ComputeOverlappedKernel(const int nthreads,
...
+  if (Caffe::GetDefaultDevice()->backend() == BACKEND_CUDA) {
+#ifdef USE_CUDA
+    // NOLINT_NEXT_LINE(whitespace/operators)
+    ComputeOverlappedKernel<Dtype><<<CAFFE_GET_BLOCKS(nthreads),
+        CAFFE_CUDA_NUM_THREADS>>>(nthreads, bbox_data, num_bboxes, num_classes,
+        overlap_threshold, overlapped_data);
+    CUDA_POST_KERNEL_CHECK;
+#endif //USE_CUDA
+  } else {
+#ifdef USE_GREENTEA
```
Comment (gongzg) [inline src/caffe/util/bbox_util.cu]: Please add a NOT_IMPLEMENTED stub in this type of code block.

File: src/caffe/util/bbox_util.cu [hunk 4]
```diff
@@ -415,16 +480,24 @@ __global__ void ComputeOverlappedByIdxKernel(const int nthreads,
...
+  if (Caffe::GetDefaultDevice()->backend() == BACKEND_CUDA) {
+#ifdef USE_CUDA
+    // NOLINT_NEXT_LINE(whitespace/operators)
+    ComputeOverlappedByIdxKernel<Dtype><<<CAFFE_GET_BLOCKS(nthreads),
+        CAFFE_CUDA_NUM_THREADS>>>(nthreads, bbox_data, overlap_threshold,
+        idx, num_idx, overlapped_data);
+    CUDA_POST_KERNEL_CHECK;
+#endif //USE_CUDA
+  } else {
+#ifdef USE_GREENTEA
```
Comment (gongzg) [inline src/caffe/util/bbox_util.cu]: Please add a NOT_IMPLEMENTED stub in this type of code block.

**Comment by listenlink** [comment]
Have added NOT_IMPLEMENTED stub code.

**Pull request merged**